### PR TITLE
docs: Clarify the commitment levels based on questions

### PR DESCRIPTION
**Pull request by joncinque**
#### Problem

Many people have asked about what commitment levels mean, and which to
choose.  This update includes some of the language at
`sdk/src/commitment_config.rs` and a recommendation for different use
cases.

Additionally, the preflight commitment documentation was out of date,
specifying that "max" was always used, and this is no longer the case.

#### Summary of Changes

Update the docs

### CODE REVIEW

File: docs/src/developing/clients/jsonrpc-api.md [hunk 1]
```diff
@@ -122,18 +122,33 @@ Requests can be sent in batches by sending an array of JSON-RPC request objects
...
   - It does not count votes on descendants of a block, only direct votes on that block.
   - This confirmation level also upholds "optimistic confirmation" guarantees in
     release 1.3 and onwards.
-- `"recent"` - the node will query its most recent block.
+- `"recent"` - the node will query its most recent block.  Note that the state at
+this slot is not derived from a finalized block.
+
+For processing many dependent transactions in series, it's recommended to use
+`"singleGossip"` commitment, which balances speed with rollback saftey.
+For total safety, it's recommended to use`"max"` commitment.
```
Comment (joncinque) [inline docs/src/developing/clients/jsonrpc-api.md:149]: What do you think about putting in a recommendation?  We've received a lot of questions about it in the past and usually we say `singleGossip`

File: docs/src/developing/clients/jsonrpc-api.md [hunk 2]
```diff
@@ -2668,9 +2683,10 @@ Submits a signed transaction to the cluster for processing.
 Before submitting, the following preflight checks are performed:
 
 1. The transaction signatures are verified
-2. The transaction is simulated against the latest max confirmed bank
```
Comment (joncinque) [inline docs/src/developing/clients/jsonrpc-api.md:2671]: The main update is to get rid of "max" here

File: docs/src/developing/clients/jsonrpc-api.md [hunk 3]
```diff
@@ -2668,9 +2683,10 @@ Submits a signed transaction to the cluster for processing.
...
-   and on failure an error will be returned. Preflight checks may be disabled if
-   desired.
+2. The transaction is simulated against the bank slot specified by the preflight
+   commitment, and on failure an error will be returned. Preflight checks may be
+   disabled if desired. It is recommended to specify the same commitment and
+   preflight commitment to avoid confusing behavior.
```
Comment (joncinque) [inline docs/src/developing/clients/jsonrpc-api.md:2689]: We don't have checks for this, which makes me think there is a use case for different levels of commitment on simulation and processing.  Perhaps this gets updated as well in web3 and rpc client?

**Comment by joncinque** [comment]
@jackcmay for :eyes: since you've been making changes around here.  Also, do docs PRs require backporting?

**Comment by jackcmay** [comment]
Yes they do, I think well do a wholesale doc jackpot once we iterate and
are happy with the new docs

**Comment by joncinque** [comment]
@CriesofCarrots you've done a lot of work and answering questions in this area -- could you take a look at this doc update?

File: docs/src/developing/clients/jsonrpc-api.md [hunk 4]
```diff
@@ -122,18 +122,33 @@ Requests can be sent in batches by sending an array of JSON-RPC request objects
...
 
-Solana nodes choose which bank state to query based on a commitment requirement
-set by the client. Clients may specify either:
-
-- `"max"` - the node will query the most recent block confirmed by supermajority of the cluster as having reached
-maximum lockout.
-- `"root"` - the node will query the most recent block having reached maximum lockout on this node.
+For preflight checks and transaction processing, Solana nodes choose which bank
+state to query based on a commitment requirement set by the client. The
+commitment describes how finalized a block is at some point in time.  When
```
Comment (CriesofCarrots) [inline docs/src/developing/clients/jsonrpc-api.md:127]:
```suggestion
commitment describes how finalized a block is at that point in time.  When
```

File: docs/src/developing/clients/jsonrpc-api.md (hunk 1 above)
Comment (CriesofCarrots) [reply docs/src/developing/clients/jsonrpc-api.md:149]: I like it

File: docs/src/developing/clients/jsonrpc-api.md [hunk 5]
```diff
@@ -2668,9 +2683,10 @@ Submits a signed transaction to the cluster for processing.
 Before submitting, the following preflight checks are performed:
 
 1. The transaction signatures are verified
-2. The transaction is simulated against the latest max confirmed bank
-   and on failure an error will be returned. Preflight checks may be disabled if
-   desired.
+2. The transaction is simulated against the bank slot specified by the preflight
+   commitment, and on failure an error will be returned. Preflight checks may be
```
Comment (CriesofCarrots) [inline docs/src/developing/clients/jsonrpc-api.md:2687]:
```suggestion
   commitment. On failure an error will be returned. Preflight checks may be
```

File: docs/src/developing/clients/jsonrpc-api.md (hunk 3 above)
Comment (CriesofCarrots) [reply docs/src/developing/clients/jsonrpc-api.md:2689]: More that in the past, it wasn't possible to tune the commitment of the simulation. I haven't thought of a strong use case for simulating and processing at different commitments.
There is client probably work we can do to make this smoother/less confusing.

File: docs/src/developing/clients/jsonrpc-api.md [hunk 6]
```diff
@@ -122,18 +122,33 @@ Requests can be sent in batches by sending an array of JSON-RPC request objects
...
+- `"root"` - the node will query the most recent block having reached maximum
+lockout on this node, meaning the node has recognized this block as finalized
 - `"singleGossip"` - the node will query the most recent block that has been voted on by supermajority of the cluster.
   - It incorporates votes from gossip and replay.
   - It does not count votes on descendants of a block, only direct votes on that block.
   - This confirmation level also upholds "optimistic confirmation" guarantees in
     release 1.3 and onwards.
-- `"recent"` - the node will query its most recent block.
+- `"recent"` - the node will query its most recent block.  Note that the state at
+this slot is not derived from a finalized block.
```
Comment (CriesofCarrots) [inline docs/src/developing/clients/jsonrpc-api.md:145]:
```suggestion
- `"recent"` - the node will query its most recent block.  Note that the block
may not be complete.
```

I think this is what you mean, right?
Finalized means reached max lockout. While it's also true that recent banks have not been finalized, I don't think that's the distinction you were making. Do I have that right?

**Review by CriesofCarrots — DISMISSED** [review]
Lgtm! r+ one change to the description of `recent`, and a couple nits for readability

File: docs/src/developing/clients/jsonrpc-api.md (hunk 6 above)
Comment (joncinque) [reply docs/src/developing/clients/jsonrpc-api.md:145]: That's much clearer, yes

**Pull request merged**